data/src/main/java/com/linkedin/data/NonBlockingDataParser.java

```diff
@@ -0,0 +1,184 @@
+/*
+   Copyright (c) 2020 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.data;
+
+import java.io.IOException;
+
+
+/**
+ * Data parser interface invoked by non blocking decoder.
+ *
+ * This interface contains methods that are invoked when parsing a Data object.
+ * Each method represents a different kind of event/read action
+ *
+ * Methods can throw IOException as a checked exception to indicate parsing error.
+ *
+ * @author amgupta1
+ */
+public interface NonBlockingDataParser
+{
+  /**
+   * Internal tokens, used to identify types of elements in data during decoding
+   */
+  enum Token
+  {
+    /**
+     * START_OBJECT is returned when encountering signals starting of an Object/map value.
+     */
+    START_OBJECT,
+    /**
+     * END_OBJECT is returned when encountering signals ending of an Object/map value
+     */
+    END_OBJECT,
+    /**
+     * START_ARRAY is returned when encountering signals starting of an Array value
+     */
+    START_ARRAY,
+    /**
+     * END_ARRAY is returned when encountering signals ending of an Array value
+     */
+    END_ARRAY,
+    /**
+     * STRING is returned when encountering a string value, field name or reference
+     */
+    STRING,
+    /**
+     * RAW_BYTES is returned when encountering chunk of raw bytes
+     */
+    RAW_BYTES,
+    /**
+     * INTEGER is returned when encountering integer value
+     */
+    INTEGER,
+    /**
+     * LONG is returned when encountering long value
+     */
+    LONG,
+    /**
+     * FLOAT is returned when encountering float decimal value
+     */
+    FLOAT,
+    /**
+     * DOUBLE is returned when encountering double decimal value
+     */
+    DOUBLE,
+    /**
+     * BOOL_TRUE is returned when encountering boolean true value
+     */
+    BOOL_TRUE,
+    /**
+     * BOOL_FALSE is returned when encountering boolean false value
+     */
+    BOOL_FALSE,
+    /**
+     * NULL is returned when encountering "null" in value context
+     */
+    NULL,
+    /**
+     * NOT_AVAILABLE can be returned if {@link NonBlockingDataParser} implementation can not currently
+     * return the requested token (usually next one), or even if any will be available,
+     * but that may be able to determine this in future. This is the case with non-blocking parsers --
```

Comment:
the "or" part is not clear to me. can you explain the usecase?

Reply:
I can remove it, it was just left it there to put more focus on first part that even if it will available but is put in buffer.
Ex. Encountering start object ordinal but not returning it till parsing the size in protobuf.

Reply:
I think it is okay to remove that part as that is implementation detail. Just document that the parser can return this if it doesn't have enough data to return the next token.